File: src/CSharp/CodeCracker/Usage/UnusedParametersCodeFixProvider.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CodeActions;
using Microsoft.CodeAnalysis.CodeFixes;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.FindSymbols;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Composition;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace CodeCracker.CSharp.Usage
{
    [ExportCodeFixProvider(LanguageNames.CSharp, Name = nameof(UnusedParametersCodeFixProvider)), Shared]
    public class UnusedParametersCodeFixProvider : CodeFixProvider
    {
        public sealed override ImmutableArray<string> FixableDiagnosticIds =>
            ImmutableArray.Create(DiagnosticId.UnusedParameters.ToDiagnosticId());

        public sealed override FixAllProvider GetFixAllProvider() => UnusedParametersCodeFixAllProvider.Instance;

        public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
        {
            var diagnostic = context.Diagnostics.First();
            context.RegisterCodeFix(CodeAction.Create(
                $"Remove unused parameter: '{diagnostic.Properties["identifier"]}'", c => RemoveParameterAsync(context.Document, diagnostic, c), nameof(UnusedParametersCodeFixProvider)), diagnostic);
            return Task.FromResult(0);
        }

        public async static Task<Solution> RemoveParameterAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
        {
            var solution = document.Project.Solution;
            var newSolution = solution;
            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
            var parameter = root.FindToken(diagnostic.Location.SourceSpan.Start).Parent.AncestorsAndSelf().OfType<ParameterSyntax>().First();
            var docs = await RemoveParameterAsync(document, parameter, root, cancellationToken);
            foreach (var doc in docs)
                newSolution = newSolution.WithDocumentSyntaxRoot(doc.DocumentId, doc.Root);
            return newSolution;
        }

        public async static Task<List<DocumentIdAndRoot>> RemoveParameterAsync(Document document, ParameterSyntax parameter, SyntaxNode root, CancellationToken cancellationToken)
        {
            var solution = document.Project.Solution;
            var parameterList = (ParameterListSyntax)parameter.Parent;
            var parameterPosition = parameterList.Parameters.IndexOf(parameter);
            var newParameterList = parameterList.WithParameters(parameterList.Parameters.Remove(parameter));
            var foundDocument = false;
            var semanticModel = await document.GetSemanticModelAsync(cancellationToken);
            var method = (BaseMethodDeclarationSyntax)parameter.Parent.Parent;
            var methodSymbol = semanticModel.GetDeclaredSymbol(method);
            var references = await SymbolFinder.FindReferencesAsync(methodSymbol, solution, cancellationToken).ConfigureAwait(false);
            var documentGroups = references.SelectMany(r => r.Locations).GroupBy(loc => loc.Document);
            var docs = new List<DocumentIdAndRoot>();
            foreach (var documentGroup in documentGroups)
            {
                var referencingDocument = documentGroup.Key;
                SyntaxNode locRoot;
                SemanticModel locSemanticModel;
                var replacingArgs = new Dictionary<SyntaxNode, SyntaxNode>();
                if (referencingDocument.Equals(document))
                {
                    locSemanticModel = semanticModel;
                    locRoot = root;
                    replacingArgs.Add(parameterList, newParameterList);
                    foundDocument = true;
                }
                else
                {
                    locSemanticModel = await referencingDocument.GetSemanticModelAsync(cancellationToken);
                    locRoot = await locSemanticModel.SyntaxTree.GetRootAsync(cancellationToken);
                }
                foreach (var loc in documentGroup)
                {
                    var methodIdentifier = locRoot.FindNode(loc.Location.SourceSpan);
                    var objectCreation = methodIdentifier.Parent as ObjectCreationExpressionSyntax;
                    var arguments = objectCreation != null
                        ? objectCreation.ArgumentList
                        : methodIdentifier.FirstAncestorOfType<InvocationExpressionSyntax>().ArgumentList;
                    if (parameter.Modifiers.Any(m => m.IsKind(SyntaxKind.ParamsKeyword)))
                    {
                        var newArguments = arguments;
                        while (newArguments.Arguments.Count > parameterPosition)
                        {
                            newArguments = newArguments.WithArguments(newArguments.Arguments.RemoveAt(parameterPosition));
                        }
                        replacingArgs.Add(arguments, newArguments);
                    }
                    else
                    {
                        var newArguments = arguments.WithArguments(arguments.Arguments.RemoveAt(parameterPosition));
                        replacingArgs.Add(arguments, newArguments);
                    }
                }
                var newLocRoot = locRoot.ReplaceNodes(replacingArgs.Keys, (original, rewritten) => replacingArgs[original]);
                docs.Add(new DocumentIdAndRoot { DocumentId = referencingDocument.Id, Root = newLocRoot });
            }
            if (!foundDocument)
            {
                var newRoot = root.ReplaceNode(parameterList, newParameterList);
                var newDocument = document.WithSyntaxRoot(newRoot);
                docs.Add(new DocumentIdAndRoot { DocumentId = document.Id, Root = newRoot });
            }
            return docs;
        }
        public struct DocumentIdAndRoot
        {
            internal DocumentId DocumentId;
            internal SyntaxNode Root;
        }
    }
}

File: src/VisualBasic/CodeCracker/Usage/UnusedParametersCodeFixProvider.vb
Imports System.Collections.Immutable
Imports System.Threading
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.CodeActions
Imports Microsoft.CodeAnalysis.CodeFixes
Imports Microsoft.CodeAnalysis.FindSymbols
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Usage
    <ExportCodeFixProvider(LanguageNames.VisualBasic, Name:=NameOf(UnusedParametersCodeFixProvider)), Composition.Shared>
    Public Class UnusedParametersCodeFixProvider
        Inherits CodeFixProvider

        Public Overrides Function RegisterCodeFixesAsync(context As CodeFixContext) As Task
            Dim Diagnostic = context.Diagnostics.First()
            context.RegisterCodeFix(CodeAction.Create(
                                    String.Format("Remove unused parameter: '{0}'", Diagnostic.Properties("identifier")),
                                    Function(c) RemoveParameterAsync(context.Document, Diagnostic, c),
                                    NameOf(UnusedParametersCodeFixProvider)),
                                    Diagnostic)
            Return Task.FromResult(0)
        End Function

        Public NotOverridable Overrides ReadOnly Property FixableDiagnosticIds As ImmutableArray(Of String) = ImmutableArray.Create(DiagnosticId.UnusedParameters.ToDiagnosticId())

        Public Overrides Function GetFixAllProvider() As FixAllProvider
            Return UnusedParametersCodeFixAllProvider.Instance
        End Function
        Private Shared Async Function RemoveParameterAsync(document As Document, diagnostic As Diagnostic, cancellationToken As CancellationToken) As Task(Of Solution)
            Dim solution = document.Project.Solution
            Dim newSolution = solution
            Dim root = Await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(False)
            Dim parameter = root.FindToken(diagnostic.Location.SourceSpan.Start).Parent.FirstAncestorOrSelf(Of ParameterSyntax)
            Dim docs = Await RemoveParameterAsync(document, parameter, root, cancellationToken)
            For Each doc In docs
                newSolution = newSolution.WithDocumentSyntaxRoot(doc.DocumentId, doc.Root)
            Next
            Return newSolution
        End Function

        Public Shared Async Function RemoveParameterAsync(document As Document, parameter As ParameterSyntax, root As SyntaxNode, cancellationToken As CancellationToken) As Task(Of List(Of DocumentIdAndRoot))
            Dim solution = document.Project.Solution
            Dim parameterList = DirectCast(parameter.Parent, ParameterListSyntax)
            Dim parameterPosition = parameterList.Parameters.IndexOf(parameter)
            Dim newParameterList = parameterList.WithParameters(parameterList.Parameters.Remove(parameter))
            Dim foundDocument = False
            Dim semanticModel = Await document.GetSemanticModelAsync(cancellationToken)
            Dim method = parameter.FirstAncestorOfType(GetType(SubNewStatementSyntax), GetType(MethodBlockSyntax))
            Dim methodSymbol = semanticModel.GetDeclaredSymbol(method)
            Dim references = Await SymbolFinder.FindReferencesAsync(methodSymbol, solution, cancellationToken).ConfigureAwait(False)
            Dim documentGroups = references.SelectMany(Function(r) r.Locations).GroupBy(Function(loc) loc.Document)
            Dim docs = New List(Of DocumentIdAndRoot)
            For Each documentGroup In documentGroups
                Dim referencingDocument = documentGroup.Key
                Dim locRoot As SyntaxNode
                Dim locSemanticModel As SemanticModel
                Dim replacingArgs = New Dictionary(Of SyntaxNode, SyntaxNode)
                If referencingDocument.Equals(document) Then
                    locSemanticModel = semanticModel
                    locRoot = root
                    replacingArgs.Add(parameterList, newParameterList)
                    foundDocument = True
                Else
                    locSemanticModel = Await referencingDocument.GetSemanticModelAsync(cancellationToken)
                    locRoot = Await locSemanticModel.SyntaxTree.GetRootAsync(cancellationToken)
                End If
                For Each loc In documentGroup
                    Dim methodIdentifier = locRoot.FindNode(loc.Location.SourceSpan)
                    Dim objectCreation = TryCast(methodIdentifier.Parent, ObjectCreationExpressionSyntax)
                    Dim arguments = If(objectCreation IsNot Nothing,
                        objectCreation.ArgumentList,
                        methodIdentifier.FirstAncestorOfType(Of InvocationExpressionSyntax).ArgumentList)
                    If parameter.Modifiers.Any(Function(m) m.IsKind(SyntaxKind.ParamArrayKeyword)) Then
                        Dim newArguments = arguments
                        While newArguments.Arguments.Count > parameterPosition
                            newArguments = newArguments.WithArguments(newArguments.Arguments.RemoveAt(parameterPosition))
                        End While
                        replacingArgs.Add(arguments, newArguments)
                    Else
                        Dim newArguments = arguments.WithArguments(arguments.Arguments.RemoveAt(parameterPosition))
                        replacingArgs.Add(arguments, newArguments)
                    End If
                Next
                Dim newLocRoot = locRoot.ReplaceNodes(replacingArgs.Keys, Function(original, rewritten) replacingArgs(original))
                docs.Add(New DocumentIdAndRoot With {.DocumentId = referencingDocument.Id, .Root = newLocRoot})
            Next
            If Not foundDocument Then
                Dim newRoot = root.ReplaceNode(parameterList, newParameterList)
                Dim newDocument = document.WithSyntaxRoot(newRoot)
                docs.Add(New DocumentIdAndRoot With {.DocumentId = document.Id, .Root = newRoot})
            End If
            Return docs
        End Function

        Public Structure DocumentIdAndRoot
            Friend DocumentId As DocumentId
            Friend Root As SyntaxNode
        End Structure
    End Class
End Namespace